packages/util/src/errors.ts

```diff
@@ -0,0 +1,100 @@
+enum ErrorCode {
+  INVALID_PARAM = 'INVALID_PARAM',
+  UNKNOWN_ERROR = 'UNKNOWN_ERROR',
+}
+
+interface GeneralError<T extends ErrorCode = ErrorCode> extends Partial<Error> {
+  code?: T
+}
+
+interface InvalidParamError extends GeneralError<ErrorCode.INVALID_PARAM> {
+  param?: string
+}
+
+interface UnknownError extends GeneralError<ErrorCode.UNKNOWN_ERROR> {
+  [key: string]: any
+}
+
+// Convert an ErrorCode into its Typed Error
+type CodedGeneralError<T> = T extends ErrorCode.INVALID_PARAM
+  ? InvalidParamError
+  : T extends ErrorCode.UNKNOWN_ERROR
+  ? UnknownError
+  : never
+
+function isError<K extends ErrorCode, T extends CodedGeneralError<K>>(
+  error: GeneralError,
+  code: K
+): error is T {
+  if (error && (<GeneralError>error).code === code) {
+    return true
+  }
+  return false
+}
```

Comment:
After reviewing Lodestar's approach, I quite like it and I feel like it would provide similar benefits to the approach used here but with simpler code.

Reply:
@gabrocheleau ok, nice, I've some additional comments I might be able to do later the day or tomorrow, then we should be ready to rework this and get to some final state to be applied throughout the code base! 🙂

Reply:
When "tomorrow" gets to 16 days, lol 😬 😜.

Anyhow.

Yes, I think this Lodestar approach is going very much into a good direction, I am also feeling a lot more confident on something like this in favor to defining our own custom "throwSomething()" function, remaining on dealing with original "Errors" or sub classes seems to be a lot more robust and side effects free to me.

So I would roughly suggest that:

1. We do the generic code - similar to [errors.ts](https://github.com/ChainSafe/lodestar/blob/master/packages/utils/src/errors.ts) in a dedicated `Utils` package, as we already planned on the alternative layout

I think it then makes sense to define the respective errors like in the Lodestar [errors](https://github.com/ChainSafe/lodestar/tree/master/packages/lodestar/src/chain/errors) folder on a "per package" basis on our side - e.g. for the `block` package - I don't see too much overlap on this between the packages. This simplifies things for structuring and drawing lines (and we have some few redundancy here it also wouldn't hurt).

Additionally/eventually we can also define a few generic errors in Util as well, but rather as a second step if really necessary.

Just in between: here is a [search](https://github.com/ChainSafe/lodestar/search?q=PARENT_UNKNOWN) for the Lodestar `BlockErrorCode.PARENT_UNKNOWN` error, just to get a glimpse on error usage.

If I get this right there are no "prose error strings" (like 'The parent hash of the uncle header is not part of the canonical chain') in Lodestar but just these capital `AGGREGATOR_INDEX_TOO_HIGH = "ATTESTATION_ERROR_AGGREGATOR_INDEX_TOO_HIGH"` aggregated error strings.

I would like to keep/stick to our prose approach, already for the reason that we have all this already written down. So I guess this would be one of the major deviations from the Lodestar approach.

WDYT? This is just a first round proposal - this would again need some draft implementation to verify - and is generally open for further discussion and evolution.

Reply:
i agree the more user friendly(readable) error strings are beneficial. :+1: